### PR TITLE
Vapor 4 support + custom cache keys

### PR DESCRIPTION
**Pull request by madsodgaard**
This PR adds Vapor 4 support with the following API:
```swift
app.gatekeeper.config = .init(maxRequests: 10, per: .second)
```

Other than that it's possible to use a custom cache, but by default it uses the Vapor 4 in memory cache.
```swift
app.gatekeeper.caches.use(.memory)
```

This PR also adds `GatekeeperKeyMaker` which is a protocol used to define the cache key for a specific `Request` this solves: #3 
By default it uses `GatekeeperHostnameKeyMaker` which uses the request hostname to generate the cache key
```swift
app.gatekeeper.keyMakers.use(.hostname)
```

I still need to update the README, but the code should be ready for review.

### CODE REVIEW

**Comment by siemensikkema** [comment]
@madsodgaard Thank you for this well-crafted PR! I think the translation to Vapor 4 is very well done and the added customizability is more than welcome!
I checked out the code and made some small fixes in this branch https://github.com/nodes-vapor/gatekeeper/tree/feature/vapor4-PR-review. I find doing PR-reviews on such large changes a bit cumbersome. If you agree with the changes you can just merge that branch in (or cherry-pick those commits you agree with). With that the only thing really missing should be the README. Do you mind updating that to show the new syntax and features? 🙏🏻

**Comment by madsodgaard** [comment]
@siemensikkema Great to hear! I'll get that branch merged and update the README within a few days.

**Comment by madsodgaard** [comment]
@siemensikkema Merged your changes and updated the README. Should be ready for review!

File: README.md [hunk 1]
```diff
@@ -68,11 +57,62 @@ protectedRoutes.get("protected/hello") { req in
...
 ```swift
 // Register middleware
-var middlewares = MiddlewareConfig() // Create _empty_ middleware config
-middlewares.use(GatekeeperMiddleware.self)
-services.register(middlewares)
+app.middlewares.use(GatekeeperMiddleware())
+```
+
+#### Customizing config
+By default `GatekeeperMiddleware` uses `app.gatekeeper.config` as it's configuration.
```
Comment (siemensikkema) [inline README.md]:
```suggestion
By default `GatekeeperMiddleware` uses `app.gatekeeper.config` as its configuration.
```

File: README.md [hunk 2]
```diff
@@ -68,11 +57,62 @@ protectedRoutes.get("protected/hello") { req in
...
+}
+```
+**configure.swift:**
+```swift
+app.gatekeeper.keyMakers.use(.userID)
 ```
 
+## Cache 🗄
+Gatekeeper uses the same cache as configured by `app.caches.use()` from Vapor, by default.
+Therefore it is **important** to setup Vapor's cache if you're using this default behaviour. You can use an in-memory cache for Vapor like so:
```
Comment (siemensikkema) [inline README.md]:
```suggestion
Therefore it is **important** to set up Vapor's cache if you're using this default behaviour. You can use an in-memory cache for Vapor like so:
```

File: README.md [hunk 3]
```diff
@@ -68,11 +57,62 @@ protectedRoutes.get("protected/hello") { req in
...
+app.cache.use(.memory)
+```
+
+### Custom cache
+You can override which cache to use by creating your own type that conforms to the `Cache` protocol from Vapor. Use `app.gatekeeper.caches.use()` to configure which cache to use.
+
+
 ## Credits 🏆
 
 This package is developed and maintained by the Vapor team at [Nodes](https://www.nodesagency.com).
```
Comment (siemensikkema) [inline README.md:118]: It feels wrong for you not to get any credit here ☺️ 

```suggestion
This package is developed and maintained by the Vapor team at [Nodes](https://www.nodesagency.com). Special thanks goes to @madsodgaard for his work on the Vapor 4 version!
```

**Review by siemensikkema — COMMENTED** [review]
Thanks for updating. We're nearly there! Just a few nits in relation to the README and I discovered that `.swiftpm` has been committed. Could you remove that and add `.swiftpm` to `.gitignore`?

**Comment by madsodgaard** [comment]
@siemensikkema Done 😄

**Review by siemensikkema — APPROVED** [review]
Thank you so much! I'll release this in a minute 👌🏻

**Pull request merged**